File: tools/search_posts.py
#!/usr/bin/env python3
"""Search Bluesky posts. Prints one indexed line per post — see
post_index.print_indexed for the exact format and how to reference a post.

Takes one or more terms; each term may be several words. Multi-word terms are
matched as exact phrases, and results must contain every term:

    python3 search_posts.py "solar punk" --limit 10
    python3 search_posts.py "mechanical design" "open source" --limit 10
"""
import argparse
import sys

from bsky import BskyError, get
from post_index import print_indexed


def build_query(terms: list[str]) -> str:
    """Combine terms into a Bluesky `q` string: multi-word terms become exact
    phrases (quoted), and space-joining them requires all terms to appear."""
    parts = []
    for term in terms:
        term = term.strip().strip('"')
        if not term:
            continue
        parts.append(f'"{term}"' if " " in term else term)
    return " ".join(parts)


def search_posts(query: str, limit: int) -> list[dict]:
    return get("app.bsky.feed.searchPosts", {"q": query, "limit": limit})["posts"]


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("terms", nargs="+", help="one or more search terms (each may be multi-word)")
    parser.add_argument("--limit", type=int, default=10)
    args = parser.parse_args()

    query = build_query(args.terms)
    # Name the terms back on an empty result, and say why it can be empty:
    # results must contain EVERY term, so a list of related topics matches
    # almost nothing. Without this the model sees a blank response and guesses.
    empty = (
        f"(no posts contain all of: {', '.join(args.terms)} — "
        "every term must appear in the same post, so search one topic at a time)"
        if len(args.terms) > 1
        else f"(no posts matched: {args.terms[0]})"
    )
    print_indexed(search_posts(query, args.limit), empty=empty)


if __name__ == "__main__":
    try:
        main()
    except BskyError as e:
        print(f"search failed: {e}", file=sys.stderr)
        sys.exit(1)

File: agent/.opencode/tools/search-posts.ts
import { tool } from "@opencode-ai/plugin"
import { $ } from "bun"

// Thin wrapper around tools/search_posts.py — located via the git worktree so
// there are no hardcoded paths. The Python side handles Bluesky auth, the
// Cloudflare-friendly curl transport, and the stable [N] index assignment.
export default tool({
  description:
    "Search Bluesky posts. Terms are combined with AND, NOT or — a post must " +
    "contain EVERY term you pass, so a list of related topics matches almost " +
    "nothing. Search one topic at a time and call the tool again for the next " +
    'one. Use several terms only to narrow a single topic, e.g. terms: ' +
    '["mechanical design", "open source"] to find posts about both at once. ' +
    "Multi-word terms are matched as exact phrases. Returns one line per " +
    "post: `[N] @handle (K likes) " +
    "at://uri — preview`. To show a post in your reply, write just its " +
    "bracketed index like [N]; the UI renders it as a live card.",
  args: {
    terms: tool.schema
      .array(tool.schema.string())
      .describe('Search terms, each may be multi-word, e.g. ["mechanical design", "open source"]'),
    limit: tool.schema.number().describe("Max posts to return").default(10),
  },
  async execute(args, context) {
    const script = `${context.worktree}/tools/search_posts.py`
    // Guard limit — the model sometimes omits it and a schema default doesn't
    // always reach here, which would send the literal "undefined" (argparse
    // exit 2).
    const limit = Number.isFinite(args.limit) ? Math.trunc(args.limit as number) : 10
    const terms = (Array.isArray(args.terms) ? args.terms : [String(args.terms ?? "")])
      .map((t) => String(t).trim())
      .filter(Boolean)
    // Each term is passed as its own argv entry (Bun escapes array elements
    // individually), so multi-word terms survive intact and Python re-quotes
    // them into exact phrases.
    return await $`python3 ${script} ${terms} --limit ${String(limit)}`.env({ ...process.env, ACTING_SESSION: context.sessionID }).text()
  },
})
